### DISCUSSION TITLE
About eclipse product debugging with maven tycho

### DISCUSSION THREAD

**Discussion by vimuxx** [eclipse-tycho/tycho#1284] [Q&A] [answered]
Hello, Guys

Sorry for disturbing.

I'm now trying to find a way to debug eclipse product with maven tycho. After
searching and experimenting for weeks, still not found the way out, please help!

Don't know why, when I use the way of PDE build, it always encounter some
inexplicable build errors. Sometimes it seems that the target platform is not
fully parsed but no errors are reported. And I always need to do extra works
when i export product from .product file, cause of some plugins are not being
exported to my export directory.

Maven does all this work very good, it's simple, no errors and no extra works
needed. So i want to giveup on the old PDE way and completely enjoy working with
maven tycho. And here is the last problem:

How to debug whole eclipse product with maven tycho in the right way?

What i have tried:

1) Launch an Eclipse application in Debug from the .product file

    With my eclipse IDE product

    Seems that this execution could start building the product only in the way
    of PDE build but not a Maven build.

2) Debug Configurations - Maven Build

    With my eclipse IDE product

    Fill the Goals with: package
    It built the whole product, nut no product start running.

3) Debug Configurations - Maven Build

    With a simple hello world application generated by mvn archetype:generate.

    Fill the Goals with: exec:exec -Dexec.executable=java -Dexec.args="-cp
    target/zxczxc-1.0-SNAPSHOT.jar org.qweqwe.asdasd.App"

    It starts the application and no breakpoints set in source file be catched.

4) Debug Configurations - Maven Build

    With my eclipse IDE product

    Fill the Goals with: exec:exec -Dexec.executable=java -Dexec.args="-cp org.qweqwe.asdasd.product/target/products/org.qweqwe.asdasd.ide/linux/gtk/x86_64/plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar org.eclipse.equinox.launcher.Main"

    It starts my product and no breakpoints set in the source file be catched.

5) Debug Configurations - Maven Build

    With a simple hello world application generated by mvn archetype:generate.

    Fill the Goals with: exec:java -Dexec.mainClass=org.qweqwe.asdasd.App

    It starts the application and could catch the breakpoints set in the file.

6) Debug Configurations - Maven Build

    With my eclipse IDE product

    Fill the Goals with: exec:java -Dexec.mainClass=org.eclipse.equinox.launcher.Main
    Not executable with output

        [WARNING]
        java.lang.ClassNotFoundException: org.eclipse.equinox.launcher.Main
            at org.codehaus.mojo.exec.URLClassLoaderBuilder$ExecJavaClassLoader.loadClass(URLClassLoaderBuilder.java:198)
            at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
            at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:271)
            at java.base/java.lang.Thread.run(Thread.java:829)

7) Maven Surefire Plugin, Tycho Surefire Plugin

    Sadly don't know how to use these to start and debug the whole IDE product.

Please give me a hand.

Appreciation!

**Comment by mickaelistria** [comment, marked as answer]
Tycho is a build tool more than a development tool. PDE remains the reference development tool and Tycho does provide any good alternative to workflow 1. described above.
What are the actual issues that make you unsatisfied with Debug As>Eclipse application? I think you should focus on getting those issues mitigated instead of trying to runaway from them.

**Reply by vimuxx** [reply]
Thanks for reply.

Debug As > Eclipse application works good after a PDE build but not a maven build that i want.

There still some build/export issues that i don't understand and don't even know how to describe them, it may takes a long time to understand and get them solved.

Few stupid questions:

Q1: You mean "Tycho does NOT provide any good alternative to workflow 1. described above.", Right?
Q2: Why workflow 6. throws ClassNotFoundException?
Q3: How to improve maven or tycho so could workflow 6. works just like workflow 5.? Is it complicated?

**Reply by laeubi** [reply]
@vimuxx what you like to do is start a remote debugging session:

https://stackoverflow.com/a/975342/9503281

then you can connect with your eclipse to the process running by `exec`

**Reply by vimuxx** [reply]
@laeubi Many thanks. It works.

**Comment by vimuxx** [comment]
mvn exec:exec -Dexec.executable=java -Dexec.args="-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:8000 -jar ./path/to/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar"
